File: lib/utils/ae2.lua
local function transferItemFromAE2(itemId, targetPeripheralType, minCount, maxCount, targetSlot)
    local interface = peripheral.find("appliedenergistics2:interface")

    if not interface then
        print("Could not find AE2 interface")
        return false
    end

    -- Find available item
    local item = interface.findItem(itemId)
    if not item then
        print("No " .. itemId .. " in AE2 system")
        return false
    end

    -- Get all peripherals
    local names = peripheral.getNames()

    -- Find all matching peripherals
    for _, destination in ipairs(names) do
        local pType = peripheral.getType(destination)
        if pType == targetPeripheralType then
            local device = peripheral.wrap(destination)

            local slot = device.getItemMeta(targetSlot)
            local currentCount = slot and slot.count or 0

            if currentCount < minCount then
                local count = maxCount - currentCount
                local metadata = item.getMetadata()
                local itemCount = metadata.count

                if itemCount >= count then
                    item.export(destination, count)
                else
                    return false
                end
            end
        end
    end
end

local function findOrCraftItem(itemId, count)
    local interface = peripheral.find("appliedenergistics2:interface")

    if not interface then
        print("Could not find AE2 interface")
        return false
    end

    -- First try to find the item
    local item = interface.findItem(itemId)
    if item then
        local metadata = item.getMetadata()
        if metadata.count >= count then
            return item
        end
    end

    -- If item not found or not enough, try to craft it
    print("Attempting to craft " .. itemId)
    item.craft(count)

    return false
end

return {
    transferItemFromAE2 = transferItemFromAE2,
    findOrCraftItem = findOrCraftItem
}

File: src/utils/ae2.ts
interface IGrid {
    findItem(item: string): IAEItemStack
    findItems(item: string): IAEItemStack[]
    getCraftingCPUs(): unknown
    getNetworkEnergyStored(): number
    getNetworkEnergyUsage(): number
    listAvailableItems(): unknown
}

interface Item {
    getMetadata(): {
        count: number
    }
}
interface IAEItemStack extends Item {
    craft(quantity: number): unknown
    export(toName: string, limit?: number, toSlot?: number): number
}

export function transferItemFromAE2(itemId: string, targetPeripheralType: string, minCount: number, maxCount: number, targetSlot: number): void {
    const interfacePeripheral = peripheral.find("appliedenergistics2:interface")[0] as unknown as IGrid;

    if (!interfacePeripheral) {
        print("Could not find AE2 interface");
        return
    }

    const item = interfacePeripheral.findItem(itemId);
    if (!item) {
        print(`No ${itemId} in AE2 system`);
        return
    }

    const names = peripheral.getNames();
    for (const destination of names) {
        const pType = peripheral.getType(destination) as unknown as string;
        if (pType === targetPeripheralType) {
            const device = peripheral.wrap(destination) as {
                getItemMeta: (slot: number) => {
                    count: number
                }
            };

            const slot = device.getItemMeta(targetSlot);
            const currentCount = slot ? slot.count : 0;

            if (currentCount < minCount) {
                const count = maxCount - currentCount;
                const metadata = item.getMetadata();
                const itemCount = metadata.count;

                if (itemCount >= count) {
                    item.export(destination, count);
                } else {
                    return;
                }
            }
        }
    }
    return;
}

export function findOrCraftItem(itemId: string, count: number): IAEItemStack | undefined {
    const interfacePeripheral = peripheral.find("appliedenergistics2:interface") as unknown as IGrid;

    if (!interfacePeripheral) {
        print("Could not find AE2 interface");
        return;
    }

    const item = interfacePeripheral.findItem(itemId);
    if (item) {
        const metadata = item.getMetadata();
        if (metadata.count >= count) {
            return item;
        }
    }

    print(`Attempting to craft ${itemId}`);
    item.craft(count);

    return;
}